### PR TITLE
Make cmake search for preinstalled protobuf packages in package mode and use officially documented variable names for getting library names

### PR DESCRIPTION
**Pull request by kskalski**
This change primarily removes the "CONFIG" suffix in find_package(protobuf), whose purpose isn't entirely clear to me (what config are we looking for and in which cases / systems is it installed with packages?).
This way protobuf 3.0 detection works properly e.g. on Debian (Raspbian).

It also adds fallback to use the proper names of variables defined by protobuf cmake package according to https://cmake.org/cmake/help/v3.1/module/FindProtobuf.html when "TARGET" conditions doesn't hold (again not entirely sure what are those targets, since they are not mentioned in official docs).

### CODE REVIEW

**Comment by kskalski** [comment]
ping

**Comment by jtattermusch** [comment]
@podsvirov , can you please comment on the intended meaning of `find_package(protobuf CONFIG)` introduced here: https://github.com/grpc/grpc/pull/7553/files#diff-14d5339fff038df1b7418aa8135bb4e8R130

CC @nicolasnoble

File: templates/CMakeLists.txt.template [hunk 1]
```diff
@@ -40,17 +40,17 @@
   # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
   <%!
-  
+
```
Comment (jtattermusch) [inline templates/CMakeLists.txt.template]: Please revert all the formatting changes (removing empty lines)

File: templates/CMakeLists.txt.template [hunk 2]
```diff
@@ -196,18 +196,24 @@
         message(WARNING "gRPC_PROTOBUF_PROVIDER is \"module\" but PROTOBUF_ROOT_DIR is wrong")
     endif()
   elseif("<%text>${gRPC_PROTOBUF_PROVIDER}</%text>" STREQUAL "package")
-    find_package(protobuf CONFIG)
-    if(protobuf_FOUND)
+    find_package(Protobuf)
+    if(Protobuf_FOUND)
```
Comment (jtattermusch) [inline templates/CMakeLists.txt.template]: Shouldn't this be PROTOBUF_FOUND in caps as for all the other packages?

File: templates/CMakeLists.txt.template [hunk 3]
```diff
@@ -196,18 +196,24 @@
         message(WARNING "gRPC_PROTOBUF_PROVIDER is \"module\" but PROTOBUF_ROOT_DIR is wrong")
     endif()
   elseif("<%text>${gRPC_PROTOBUF_PROVIDER}</%text>" STREQUAL "package")
-    find_package(protobuf CONFIG)
-    if(protobuf_FOUND)
+    find_package(Protobuf)
```
Comment (jtattermusch) [inline templates/CMakeLists.txt.template]: PROTOBUF?

**Comment by podsvirov** [comment]
Hello @kskalski and @jtattermusch.

Please read [Using Packages](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#using-packages) in CMake documentation to lear the diffirents betwen `CONFIG` ([Config-file Packages](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#config-file-packages)) and `MODULE` ([Find-module Packages](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#find-module-packages)).

Since version 3.0 protobuf can be built with CMake (officialy on Windows only, but realy on ohter platforms too) and provide `Config-file Package` named `protobuf` with [IMPORTED](https://cmake.org/cmake/help/latest/prop_tgt/IMPORTED.html#prop_tgt:IMPORTED) targets like `protobuf::libprotbuf` and `protobuf::protoc`.
Also CMake have built-in `Find-module Package` named `Protobuf`.
It's two diffirend ways to find protobuf libraries and compiler.

In the beginning, the gRPC's CMake project did not support the use of packages at all.
Then I added an optional support for using `Config-file Packages` and `IMPORTED` targets.
If you want to use `Find-module Packages`, please add support for this new functionality without damaging the existing functionality.

**Comment by kskalski** [comment]
@podsvirov "If you want to use Find-module Packages, please add support for this new functionality without damaging the existing functionality." - isn't it what my change actually does?
From documentation at https://cmake.org/cmake/help/v3.0/command/find_package.html:
"The command has two modes by which it searches for packages: “Module” mode and “Config” mode. Module mode is available when the command is invoked with the above reduced signature. CMake searches for a file called Find<package>.cmake in the CMAKE_MODULE_PATH followed by the CMake installation. If the file is found, it is read and processed by CMake. It is responsible for finding the package, checking the version, and producing any needed messages. Many find-modules provide limited or no support for versioning; check the module documentation. **If no module is found and the MODULE option is not given the command proceeds to Config mode.**"

So basically having "find_package(protobuf CONFIG)" disables the MODULE mode, while "find_package(Protobuf)" tries both. Note that I kept the checks for protobuf::protoc style targets, so there is should also work fine for both mode. Unfortunately I'm not sure how to tests it, on the Raspbian distro that I'm testing it on the only way that works is MODULE, that's why I made the change.

**Comment by podsvirov** [comment]
I use `IMPORTED` targets to make [Relocatable Packages](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#creating-relocatable-packages). It's some works [!861](https://gitlab.kitware.com/cmake/cmake/merge_requests/861) and [!876](url) in CMake. And it will be useful for me too.
In the near future I will try to provide a solution that will satisfy both of us.

**Comment by podsvirov** [comment]
Wow:

> Files changed 1,948

It's not good...

File: templates/CMakeLists.txt.template (hunk 1 above)
Comment (kskalski) [inline templates/CMakeLists.txt.template]: Ok, I rebased to upstream and the formatting changes are gone

File: templates/CMakeLists.txt.template (hunk 3 above)
Comment (kskalski) [inline templates/CMakeLists.txt.template]: hm, looks like I got warnings when doing that:
CMake Warning at CMakeLists.txt:187 (find_package):
  By not providing "FindPROTOBUF.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "PROTOBUF",
  but CMake did not find one.

  Could not find a package configuration file provided by "PROTOBUF" with any
  of the following names:

    PROTOBUFConfig.cmake
    protobuf-config.cmake

I think the proper name of the package is Protobuf, this is also what Konstantin mentioned in his comments

File: templates/CMakeLists.txt.template (hunk 2 above)
Comment (kskalski) [inline templates/CMakeLists.txt.template]: Ok, right, I'm changing to caps, it seems that those variables are also exported in such form

**Comment by kskalski** [comment]
Huh, github is not cooperating... anyway, the diff on my branch is clean https://github.com/grpc/grpc/compare/master...kskalski:protobuf_cmake , though it doesn't seem to allow creating a new PR anymore...

File: templates/CMakeLists.txt.template (hunk 1 above)
Comment (podsvirov) [inline templates/CMakeLists.txt.template]: The `Protobuf` package name сan be suitable for both package search options.
But `Find-module` package will have a permanent priority. It is advisable to add an option to configure this behavior.

Comment (kskalski) [inline templates/CMakeLists.txt.template]: Is the default order of package search that find_package is doing undesirable?
Looks like it first tries to find -module type of package and if it can't, it will proceed to look for -config type of package.

I guess the usecase that you have in mind is that there are both available on the machine, but the one of -config type is newer / more up-to-date than the other. Isn't it possible to use the  CMAKE_MODULE_PATH  to override the path it searching in or even use gRPC_PROTOBUF_PROVIDER=module (with -DPROTOBUF_ROOT_DIR=...) ?

Comment (podsvirov) [inline templates/CMakeLists.txt.template]: The `gRPC_PROTOBUF_PROVIDER=module` now mean compile and use `protobuf` from `git` source code submodule.
I think the `gRPC_PROTOBUF_PACKAGE_TYPE` with values `MODULE`, `CONFIG` or empty `` or `AUTO` to using both will be useful.

File: CMakeLists.txt [hunk 4]
```diff
@@ -66,6 +66,9 @@ set_property(CACHE gRPC_SSL_PROVIDER PROPERTY STRINGS "module" "package")
 set(gRPC_PROTOBUF_PROVIDER "module" CACHE STRING "Provider of protobuf library")
 set_property(CACHE gRPC_PROTOBUF_PROVIDER PROPERTY STRINGS "module" "package")
 
+set(gRPC_PROTOBUF_PROVIDER_TYPE "" CACHE STRING "Algorithm for searching protobuf package")
```
Comment (kskalski) [inline CMakeLists.txt]: @podsvirov Something like this?

Comment (podsvirov) [inline CMakeLists.txt]: Hm. What `MODULE` or `CONFIG` means when `gRPC_PROTOBUF_PROVIDER` is `module`?
Please use `gRPC_PROTOBUF_PACKAGE_TYPE` (as I was write above) instead.

File: CMakeLists.txt [hunk 5]
```diff
@@ -184,21 +187,24 @@ if("${gRPC_PROTOBUF_PROVIDER}" STREQUAL "module")
       message(WARNING "gRPC_PROTOBUF_PROVIDER is \"module\" but PROTOBUF_ROOT_DIR is wrong")
   endif()
 elseif("${gRPC_PROTOBUF_PROVIDER}" STREQUAL "package")
-  find_package(protobuf CONFIG)
-  if(protobuf_FOUND)
+  find_package(Protobuf ${gRPC_PROTOBUF_PROVIDER_TYPE})
+  if(PROTOBUF_FOUND)
```
Comment (podsvirov) [inline CMakeLists.txt]: Uppercase variables will be defined only from `FindProtobuf` module for backward compatibility with he old implementation (CMake less 3.6 version). This wariable will not be defined for `Config-file Package`.

Comment (podsvirov) [inline CMakeLists.txt]: Common solution `if(Protobuf_FOUND OR PROTOBUF_FOUND)` will work in all cases.

Comment (kskalski) [inline CMakeLists.txt]: Ok, done

File: CMakeLists.txt [hunk 6]
```diff
@@ -184,21 +187,24 @@ if("${gRPC_PROTOBUF_PROVIDER}" STREQUAL "module")
...
     if(TARGET protobuf::protoc)
       set(_gRPC_PROTOBUF_PROTOC protobuf::protoc)
+    else()
+      set(_gRPC_PROTOBUF_PROTOC ${PROTOBUF_PROTOC_EXECUTABLE})
     endif()
-    set(_gRPC_FIND_PROTOBUF "if(NOT protobuf_FOUND)\n  find_package(protobuf CONFIG)\nendif()")
-  else()
-    find_package(Protobuf MODULE)
-    set(_gRPC_FIND_PROTOBUF "if(NOT Protobuf_FOUND)\n  find_package(Protobuf)\nendif()")
+    set(_gRPC_FIND_PROTOBUF "if(NOT PROTOBUF_FOUND)\n  find_package(Protobuf ${gRPC_PROTOBUF_PACKAGE_TYPE})\nendif()")
```
Comment (podsvirov) [inline CMakeLists.txt]: if(NOT PROTOBUF_FOUND) -> if(NOT Protobuf_FOUND AND NOT PROTOBUF_FOUND)

**Comment by podsvirov** [comment]
I will check these changes and report the results in the near future.

File: CMakeLists.txt [hunk 7]
```diff
@@ -184,21 +187,24 @@ if("${gRPC_PROTOBUF_PROVIDER}" STREQUAL "module")
...
       set(_gRPC_PROTOBUF_PROTOC protobuf::protoc)
+    else()
+      set(_gRPC_PROTOBUF_PROTOC ${PROTOBUF_PROTOC_EXECUTABLE})
     endif()
-    set(_gRPC_FIND_PROTOBUF "if(NOT protobuf_FOUND)\n  find_package(protobuf CONFIG)\nendif()")
-  else()
-    find_package(Protobuf MODULE)
-    set(_gRPC_FIND_PROTOBUF "if(NOT Protobuf_FOUND)\n  find_package(Protobuf)\nendif()")
+    set(_gRPC_FIND_PROTOBUF "if(NOT Protobuf_FOUND AND NOT PROTOBUF_FOUND)\n  find_package(Protobuf ${gRPC_PROTOBUF_PACKAGE_TYPE})\nendif()")
   endif()
```
Comment (podsvirov) [inline CMakeLists.txt]: Please add:
```
  if(PROTOBUF_FOUND)
    include_directories(${PROTOBUF_INCLUDE_DIRS})
  endif()
```

Comment (kskalski) [inline CMakeLists.txt]: done

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by jtattermusch** [comment]
@kskalski do you want to `rebase -i upstream/master` your PR? Something is wrong with your commit history (it shows 250+ commits and 2777 files changes)  so this change is not mergeable or reviewable at the moment.  Free free close this PR and open a new one with fresh history if needed (and link to the new PR here).

**Comment by kskalski** [comment]
I created a new PR https://github.com/grpc/grpc/pull/11633
